### PR TITLE
fix: calculate new version from project instead of tags

### PR DESCRIPTION
**Pull request by tomplarge**
### Contributor checklist

- [x] My commit messages follow the [Conventional Commit Message Format](https://gist.github.com/stephenparish/9941e89d80e2bc58a153#format-of-the-commit-message)
      For example: `fix: Multiply by appropriate coefficient`, or
      `feat(Calculator): Correctly preserve history`
      Any explanation or long form information in your commit message should be
      in a separate paragraph, separated by a blank line from the primary message
- [x] Bug fixes include regression tests
- [x] Features include unit/acceptance tests

Also some clean up and more tests (mostly for edge cases)

### CODE REVIEW

File: test/version_test.exs [hunk 1]
```diff
@@ -33,86 +33,118 @@ defmodule GitOps.Test.VersionTest do
...
   end
 
   test "a release candidate raises correctly when it would not change" do
     assert_raise RuntimeError, ~r/No changes should result in a new release version./, fn ->
-      new_version(["0.1.0", "0.1.1-rc0"], [chore()], rc: true)
+      new_version("0.1.1-rc0", [chore()], rc: true)
     end
   end
 
   test "if a prefix is configured, it is ignored when searching for a tag" do
```
Comment (zachdaniel) [inline test/version_test.exs]: Since we aren't searching for a tag, this can go away.

File: lib/mix/tasks/git_ops.release.ex [hunk 2]
```diff
@@ -78,51 +83,59 @@ defmodule Mix.Tasks.GitOps.Release do
...
+          commits_for_version,
+          opts
+        )
       end
 
-    new_version = String.trim_leading(prefixed_version, prefix)
+    new_version = String.trim_leading(prefixed_new_version, prefix)
 
-    GitOps.Changelog.write(path, commits_for_changelog, current_version, prefixed_version)
+    if confirm_and_tag(repo, prefixed_new_version) do
```
Comment (zachdaniel) [inline lib/mix/tasks/git_ops.release.ex]: So confirm_and_tag will commit the changes, but this writes the changelog *after* the commit step. I imagine what we'd want to do instead is to get confirmation, and then write the changelog, commit, and tag. Unless this is supposed to be a step for revising the changelog? If it is, then we should say something like: please review the changelog, add any details you feel are necessary, and commit that with `docs: update changelog`. Or something along those lines.

File: lib/mix/tasks/git_ops.release.ex [hunk 3]
```diff
@@ -57,15 +57,20 @@ defmodule Mix.Tasks.GitOps.Release do
...
 
-    mix_project = mix_project_module.project()
-    repo = GitOps.Git.init!()
+    mix_project_module = GitOps.Config.mix_project()
+
+    mix_project =
+      if mix_project_module do
+        mix_project_module.project()
+      else
+        raise "mix_project must be configured in order to use git_ops. Please see the configuration in the README.md for an example."
```
Comment (zachdaniel) [inline lib/mix/tasks/git_ops.release.ex]: We should do this in the config module itself, just for organizations sake.

Comment (zachdaniel) [reply lib/mix/tasks/git_ops.release.ex]: And also, since it can't be an environment variable (only compile time application env), we could potentially just raise in a `before_compile` step to make it cleaner.

**Comment by zachdaniel** [comment]
This all looks great. A few pieces of feedback:

1.) This is technically a few changes all in one. Seeing as this is a library that encourages granular changes, we should probably break it up into multiple PRs or multiple commits. Ideally, we start with just using the current project version instead of the historic tags, and then add the changelog + tag + commit flow improvements in a later PR. I don't mind the general cleanup that you did just living in one of those PRs.

2.) We should add a few tests to keep the coverage from going down. Even if they have to be superficial :)

File: test/version_test.exs (hunk 1 above)
Comment (tomplarge) [reply test/version_test.exs]: Yup, thought so

**Comment by tomplarge** [comment]
Sounds good, that totally went through my brain as I was making this PR. I'll break it up.

File: lib/mix/tasks/git_ops.release.ex (hunk 2 above)
Comment (tomplarge) [reply lib/mix/tasks/git_ops.release.ex]: Yeah.. I was toying around with how this would be done. At first I was thinking of displaying the changes to the user, then perhaps we could write them to a temporary changelog file. Unsure exactly what would be best.

File: test/version_test.exs [hunk 4]
```diff
@@ -33,86 +35,154 @@ defmodule GitOps.Test.VersionTest do
...
+  end
+
+  test "last valid non rc is found correctly with prefixes" do
+    versions = ["v0.0.1", "v0.1.0-rc0", "0.1.0", "0.2.0-alpha"]
+    last_rc = Version.last_valid_non_rc_version(versions, "v")
+
+    assert last_rc == "v0.0.1"
   end
+
+  # QUESTION: is this supposed to be the functionality?
```
Comment (tomplarge) [inline test/version_test.exs]: Question

**Review by zachdaniel — COMMENTED** [review]
Do we even need to know about previous pre release versions anymore? I’m honestly struggling to remember why I went all in on git tags anymore. I could swear there was a reason, but we should theoretically be able to operate only using the current version, right? If so we don’t need this function at al.

**Comment by tomplarge** [comment]
It seems like it was to be able to determine which commits were new to this release, so we calculated the last release, but that question I have above in the testing module is weird.. it doesn't seem like it's calculating the last pre release but rather just the last release

File: test/version_test.exs (hunk 4 above)
Comment (zachdaniel) [reply test/version_test.exs]: I don't remember my initial intent with this, to be honest. I say for now we just rename the function to what it actually is, which is just `last_version_later_than`. Afterwards we should just revisit the logic for how we handle the `rc` flag. I think we'll find its much simpler than I made it :) (keep in mind I hacked this together in like two afternoons :/ )

Comment (zachdaniel) [reply test/version_test.exs]: Actually its: `latest_version_greater_than`

Comment (tomplarge) [reply test/version_test.exs]: Sounds good! Yeah, I figured there were definitely some simplifications to be made
